Add browser SDK (#22)
spheronFdn/docs: pages/sdk/browser.mdx

import { Callout } from "nextra-theme-docs";
import Navigation from "components/navigation";

# Storage Browser SDK

The Spheron Storage SDK is equipped with multi-chain storage capabilities powered by Spheron.

<Callout type="warning">
  The package is only meant for Browser environments.
</Callout>

## Installation

**Using NPM**

```
npm i @spheron/browser-upload
```

**Using Yarn**

```
yarn add @spheron/browser-upload
```

## Usage

The Browser SDK allows you to upload files directly from your browser to `IPFS`, `Filecoin` or `Arweave`. \
To use the Browser SDK, you have to set up the following:

> 1. An express server that will be used by the frontend to get the `uploadToken` for uploading the data.

Reviewer: It doesn't need to be an express server.

Lets just put `An web server`

2. A web app that uses `@spheron/browser-upload` to upload files directly from browser.

### Server

> You have to set up an express server with the `/initiate-upload` endpoint.

Reviewer: ```suggestion
You have to set up an web server with and endpoint that will be used by your Web App to fetch the token that will be used to upload. In the example we have created a `/initiate-upload` endpoint.
```

Make use of the `createSingleUploadToken` method from `@spheron/storage` package to generate the `uploadToken`.

<Callout type="info">
  **NOTE:** The `uploadToken` can only be used for a single upload and has an
  expiration of 10 minutes.
</Callout>

```js
import { SpheronClient, ProtocolEnum } from "@spheron/storage";

...

app.get("/initiate-upload", async (req, res, next) => {
  try {
    const bucketName = "example-browser-upload"; // use your preferred name
    const protocol = ProtocolEnum.IPFS; // use your preferred protocol
    const token = process.env.SPHERON_TOKEN; // add your access token in .env or paste it here

    const client = new SpheronClient({ token });

    const { uploadToken } = await client.createSingleUploadToken({
      name: bucketName,
      protocol,
    });

    res.status(200).json({
      uploadToken,
    });
  } catch (error) {
    console.error(error);
    next(error);
  }
});
```

You need to [Create an Access Token](/rest-api/#creating-an-access-token) to create the client.

<Callout type="info">
  **NOTE:** Please choose `web app` type in the dashboard while generating the
  Access Token.
</Callout>

### Client

> You have to get the `uploadToken` from the server by making a request to the `/initiate-upload` endpoint.

Reviewer: ```suggestion
To get the `token` that is used to upload data from Browser, you will first need to send a request to your Backend to create the `token`. In our example we are going to send a request to the `/initiate-upload` endpoint.
```

  expiration of 10 minutes.
</Callout>

```js
import { upload } from "@spheron/browser-upload";

...

const response = await fetch(`<BACKEND_URL>/initiate-upload`); // get the access token from server
const resJson = await response.json();
const token =  resJson.uploadToken;
const uploadResult = await upload(<FILES_YOU_WANT_TO_UPLOAD>, { token });

...
```

This flow allows you to control who can use your `uploadToken` to upload files.
> You can checkout the working example [here](https://github.com/spheronFdn/sdk/tree/main/examples/browser-upload).

## Available Methods

### Upload

Used to upload files to the specified protocol.

```js
let currentlyUploaded = 0;

const { uploadId, bucketId, protocolLink, dynamicLinks } = await upload(files, {
  token: res.uploadToken,
  onChunkUploaded: (uploadedSize, totalSize) => {
    currentlyUploaded += uploadedSize;
    console.log(`Uploaded ${currentlyUploaded} of ${totalSize} Bytes.`);
  },
});
```

Function `upload` takes two parameters:

- **files** - ( File[] ) - An array of files to be uploaded.
- **configuration:** An object with the following parameters:
  1.  `configuration.token`

>       - The upload token for the upload.

Reviewer: ```suggestion
      - The upload token which was fetched from your Backend server.
```

      uploadToken,
    });
  } catch (error) {
    console.error(error);
    next(error);
  }
});
```

You need to [Create an Access Token](/rest-api/#creating-an-access-token) to create the client.

<Callout type="info">
  **NOTE:** Please choose `web app` type in the dashboard while generating the
  Access Token.
</Callout>

Function `createSingleUploadToken` takes two parameters:

- **name** - (string) - Represents the name of the bucket on which you are uploading the data.
- **protocol** - (string) - The protocol on which the data will be uploaded. The supported protocols are [ `ARWEAVE`, `IPFS`, `FILECOIN`].

**Response**

```js
interface TokenRes {
  uploadToken: string;
}
```

### Client

You have to send a request to your server to create the `uploadToken` that will be used to upload files from the browser.
In the example, we are sending a request to the `/initiate-upload` endpoint.
Make use of the `upload` method from `@spheron/browser-upload` to upload your files.

<Callout type="info">
  **NOTE:** The `uploadToken` can only be used for a single upload and has an
  expiration of 10 minutes.
</Callout>

```js
import { upload } from "@spheron/browser-upload";

...

> const response = await fetch(`<BACKEND_URL>/initiate-upload`); // get the access token from server

Reviewer 2: ```suggestion
const response = await fetch(`<BACKEND_URL>/initiate-upload`); // get the temporary access token from server
```

let currentlyUploaded = 0;

const { uploadId, bucketId, protocolLink, dynamicLinks } = await upload(files, {

>   token: res.uploadToken,

Reviewer 2: This should be a token extracted in the line 111
```
const token = resJson.uploadToken;
```
Just do this instead of using res, which is not present. If someone copies this code, it will not work. Please check the code example multiple times, as this is one of the most imp parts of the doc. Wrong code will give a very bad experience to the devs as they completely trust the code given in the docs.
```suggestion
  token,
```

const { uploadId, bucketId, protocolLink, dynamicLinks } = await upload(files, {
  token,
  onChunkUploaded: (uploadedSize, totalSize) => {
    currentlyUploaded += uploadedSize;
    console.log(`Uploaded ${currentlyUploaded} of ${totalSize} Bytes.`);
  },
});

...
```

This flow allows you to control who can use your `uploadToken` to upload files. \
Function `upload` takes two parameters:

- **files** - ( File[] ) - An array of files to be uploaded.
- **configuration:** An object with the following parameters:
  1.  `configuration.token`
      - The upload token that was fetched from server.
  2.  `configuration.onChunkUploaded` **(Optional)**
      - callback function `(uploadedSize: number, totalSize: number) => void`. The function will be called multiple times, depending on the upload size. The function will be called each time a chunk is uploaded, with two parameters. the first one `uploadedSize` represents the size in Bytes for the uploaded chunk. The `totalSize` represents the total size of the upload in Bytes.

**Response**

```js

> interface UploadRes {
>   uploadId: string;
>   bucketId: string;
>   protocolLink: string;
>   dynamicLinks: Domain[];
> }

Reviewer: ```suggestion
{
  uploadId: string;
  bucketId: string;
  protocolLink: string;
  dynamicLinks: string[];
}
```

> > You can checkout the working example [here](https://github.com/spheronFdn/sdk/tree/main/examples/browser-upload).

Reviewer 2: put this at the starting at the end of the usage section - https://github.com/spheronFdn/docs/pull/22/files#diff-d629206430e48bc9c6975ee35ebf00f29a50750b1c941c4746e50d95a01347c5R33
